near-sdk/src/collections/tree_map.rs

```diff
@@ -0,0 +1,1681 @@
+                // n - id of a node that holds key k, p - id of an immediate parent of n
+                let (k, n, p) = self.min_at(rgt, r_id).unwrap();
+
+                self.key.insert(&r_id, &k);
+                self.key.remove(&n);
+                self.ht.remove(&n);
+
+                if self.lft.get(&p).map(|id| id == n).unwrap_or_default() {
+                    // n is on left link of p
+                    match self.rgt.get(&n) {
+                        Some(r) => {
+                            self.lft.insert(&p, &r);
+                            self.rgt.remove(&n);
+                        },
+                        None => {
+                            self.lft.remove(&p);
+                        }
+                    }
+                } else {
+                    // n is on right link of p
+                    match self.rgt.get(&n) {
+                        Some(r) => {
+                            self.rgt.insert(&p, &r);
+                            self.rgt.remove(&n);
+                        },
+                        None => {
+                            self.rgt.remove(&p);
+                        }
+                    }
+                }
+
+                let root = self.swap_with_last(n);
+
+                // removing node might have caused a imbalance - balance the tree up to the root,
+                // starting from the lowest affected key (min key from right subtree in this case)
+                self.check_balance(root, &k)
+
+            }
+        }
+    }
+
+    // Move content of node with ID `len - 1` (parent left or right link, left, right, key, height)
+    // to node with id `t`, and remove node `len - 1`.
+    fn swap_with_last(&mut self, target: u64) -> u64 {
+        let remove = self.len - 1;
+        if target == remove {
+            return self.root;
+        }
+
+        let k_opt = self.key.get(&remove);
+        if k_opt.is_none() {
+            return self.root;
+        }
+        let key = k_opt.unwrap();
+
+        // parent lookup takes O(log(N)) in worst case - can take amortized O(1) with `parent` map
+        let root = self.root;
+        let (_, parent) = self.lookup_at(root, &key).unwrap();
+
+        self.key.remove(&remove);
+        self.key.insert(&target, &key);
+
+        // update link from parent node (p-n -> p-t)
+        if remove != parent {
+            if self.lft.get(&parent).map(|link| link == remove).unwrap_or_default() {
+                self.lft.insert(&parent, &target);
+            } else {
+                self.rgt.insert(&parent, &target);
+            }
+        }
+
+        // move references from `ht`, `lft`, `rgt` from `r` to `t`.
+        for map in [&mut self.ht, &mut self.lft, &mut self.rgt].iter_mut() {
+            let val = map.remove(&remove);
+            match val {
+                Some(x) => {
+                    map.insert(&target, &x);
+                },
+                None => ()
```

Comment:
In fact we do, thanks.

Reply:
How come the tests didn't catch that?
Does `prop_avl_vs_rb` catch it if we do deletions between insertions?

Reply:
Good question, seems like this is rather rare case that requires very specific input to be caught. I will try to come up with such test that must fail without `map.remove(&target)`.

Reply:
We should improve the random test until it can catch this type of bugs:
- how does quickcheck decide the range of values? if it uses whole u32 then all insertions are distinct and deletions are no-op, need meaningful values
- how does it decide the ratio of insertions of deletions? with too many deletions the map is always empty, with almost all insertions some other case may be missed
- makes sense to test lookups between those operations

Reply:
Now I see that that test do fine, there is just no bug to catch: `remove` is unnecessary - as `target`'s height and links (max 1 link) are reset in `do_remove` before `swap_with_last` is called.

Adding comment. Sorry for confusion.